### PR TITLE
Issue #1042: Throw Exception if snippet not found in key value storage

### LINKED ISSUE #1042: Throw exception "Snippet for product ID not found" in ProductJsonService
This issue happens if products are in the search engine but not in the key-value store.
This has happened during initial setup of a project when testing different backends.
See `\LizardsAndPumpkins\Http\ContentDelivery\ProductJsonService\ProductJsonService::buildProductData()`, `json_decode($snippets[$productJsonSnippetKey], true),`, currently throws error `Fatal error:  Uncaught TypeError: json_decode() expects parameter 1 to be string, null given`.
Exception should be "Snippet for product not found in key value store".

### PR DESCRIPTION
**Pull request by Schrank**
Fixes #1042

### CODE REVIEW

File: src/Http/ContentDelivery/ProductJsonService/ProductJsonService.php [hunk 1]
```diff
@@ -121,16 +122,22 @@ private function buildProductData(
         array $productJsonSnippetKeys,
         array $priceSnippetKeys,
         array $specialPriceSnippetKeys
-    ) : array {
+    ): array {
         $snippets = $this->getSnippets($productJsonSnippetKeys, $priceSnippetKeys, $specialPriceSnippetKeys);
 
         return array_map(function ($productJsonSnippetKey, $priceKey, $specialPriceKey) use ($context, $snippets) {
-            return $this->enrichProductJsonWithPrices->addPricesToProductData(
+            if ($snippets[$productJsonSnippetKey] === null) {
```
Comment (tim-bezhashvyly) [inline src/Http/ContentDelivery/ProductJsonService/ProductJsonService.php]: Yoda please.

File: src/Http/ContentDelivery/ProductJsonService/ProductJsonService.php [hunk 2]
```diff
@@ -121,16 +122,22 @@ private function buildProductData(
...
+                );
+            }
+            $productData = $this->enrichProductJsonWithPrices->addPricesToProductData(
                 $context,
                 json_decode($snippets[$productJsonSnippetKey], true),
                 $snippets[$priceKey],
-                @$snippets[$specialPriceKey]
+                $snippets[$specialPriceKey] ?? null
             );
+            return $productData;
```
Comment (tim-bezhashvyly) [inline src/Http/ContentDelivery/ProductJsonService/ProductJsonService.php]: Why not inline it?

File: tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php [hunk 3]
```diff
@@ -120,7 +123,52 @@ public function testItReturnsTheEnrichedProductData()
...
         $result = $this->productJsonService->get($stubContext, $stubProductId);
-     
+
         $this->assertContains($expected, $result);
     }
+
+    /**
+     * @param string $jsonSnippet
+     * @param string $priceSnippet
+     * @param string $specialPriceSnippet
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php]: Please use type-hints instead.

File: tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php [hunk 4]
```diff
@@ -120,7 +123,52 @@ public function testItReturnsTheEnrichedProductData()
...
+     *
+     * @dataProvider providesSnippets
+     */
+    public function testItThrowsAnExceptionIfKeyValueDoesNotContainSnippet($jsonSnippet, $priceSnippet, $specialPriceSnippet)
+    {
+        $this->expectException(ProductSnippetNotFoundInKeyValueStorageException::class);
+        $this->expectExceptionMessageRegExp('#Snippet with key .*? not found.#');
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php]: Not critical but I guess `$this->expectExceptionMessage(sprintf('Snippet with key .%s not found.', $jsonSnippetKey));` would be more exact and readable.

File: tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php [hunk 5]
```diff
@@ -120,7 +123,52 @@ public function testItReturnsTheEnrichedProductData()
...
+        $this->productJsonService->get($stubContext, $stubProductId);
+    }
+
+    public function providesSnippets()
+    {
+        return [
+            [null, '99', '89'],
+            [null, null, null],
+        ];
+    }
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php]: What is an intent of using different data sets here?

File: src/Http/ContentDelivery/ProductJsonService/ProductJsonService.php (hunk 2 above)
Comment (Schrank) [inline src/Http/ContentDelivery/ProductJsonService/ProductJsonService.php]: I think I wanted to test something on the data? no clue :)

File: tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php (hunk 5 above)
Comment (Schrank) [inline tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php]: In the beginning I wanted to test the numbers being null too, but they are not json_decoded, so it doesn't matter. So the set was there and no reason to delete it.

Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Http/ContentDelivery/ProductJsonService/ProductJsonServiceTest.php]: Of course there is. Unneeded tests are making noise.

**Comment by Schrank** [comment]
New review please :-)

File: src/Http/ContentDelivery/ProductJsonService/Exception/ProductSnippetNotFoundInKeyValueStorageException.php [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LizardsAndPumpkins\Http\ContentDelivery\ProductJsonService\Exception;
+
+class ProductSnippetNotFoundInKeyValueStorageException extends \Exception
```
Comment (tim-bezhashvyly) [inline src/Http/ContentDelivery/ProductJsonService/Exception/ProductSnippetNotFoundInKeyValueStorageException.php]: I'm wondering if `ProductSnippetNotFoundException` or even `SnippetNotFoundException` would be name? What do you think?

Comment (Schrank) [inline src/Http/ContentDelivery/ProductJsonService/Exception/ProductSnippetNotFoundInKeyValueStorageException.php]: Good idea. Shorted the name

**Pull request merged**